cache/lib/cache/authentication.ex

```diff
@@ -0,0 +1,191 @@
+defmodule Cache.Authentication do
+  @moduledoc """
+  Authentication for CAS operations using the /api/projects endpoint.
+
+  This module validates that a request has proper authorization to access a project
+  by calling the server's /api/projects endpoint and caching the results.
+  """
+
+  @failure_cache_ttl 3
+  @success_cache_ttl 600
+  @cache_name :cas_auth_cache
+
+  require Logger
+
+  def child_spec(_) do
+    %{
+      id: __MODULE__,
+      start: {Cachex, :start_link, [@cache_name, []]}
+    }
+  end
+
+  @doc """
+  Ensures the request has access to the specified project.
+
+  Returns `{:ok, auth_header}` if authorized, or `{:error, status, message}` otherwise.
+  """
+  def ensure_project_accessible(conn, account_handle, project_handle) do
+    auth_header = Plug.Conn.get_req_header(conn, "authorization") |> List.first()
+
+    Logger.info("Checking authorization for: #{account_handle}/#{project_handle} with auth header: #{auth_header}")
+
+    if is_nil(auth_header) do
+      {:error, 401, "Missing Authorization header"}
+    else
+      requested_handle = normalize_handle(account_handle, project_handle)
+
+      case authorize(auth_header, requested_handle, conn) do
+        :ok -> {:ok, auth_header}
+        {:error, status, message} -> {:error, status, message}
+      end
+    end
+  end
+
+  defp authorize(auth_header, requested_handle, conn) do
+    cache_key = generate_cache_key(auth_header)
+
+    case cached_failure(cache_key) do
+      {:error, status, message} ->
+        {:error, status, message}
+
+      :miss ->
+        case evaluate_cached_project(cache_key, requested_handle) do
+          :authorized ->
+            :ok
+
+          :unauthorized ->
+            {:error, 404, "Unauthorized or not found"}
+
+          :miss ->
+            fetch_and_cache_projects(auth_header, cache_key, requested_handle, conn)
+        end
+    end
+  end
+
+  defp cached_failure(cache_key) do
+    case Cachex.get(@cache_name, {:failure, cache_key}) do
+      {:ok, {status, message}} -> {:error, status, message}
+      _ -> :miss
+    end
+  end
+
+  defp evaluate_cached_project(cache_key, requested_handle) do
+    case Cachex.get(@cache_name, {:primed, cache_key}) do
+      {:ok, version} when is_integer(version) ->
+        case Cachex.get(@cache_name, {cache_key, requested_handle}) do
+          {:ok, {^version, :allowed}} -> :authorized
+          {:ok, nil} -> :unauthorized
+          _ -> :miss
+        end
+
+      _ ->
+        :miss
+    end
+  end
+
+  defp fetch_and_cache_projects(auth_header, cache_key, requested_handle, conn) do
+    headers = build_headers(auth_header, conn)
+    options = request_options(headers)
+
+    case Req.get(options) do
```

Comment:
Once we move to account tokens scoped to projects and a set of permissions, the assumption here will break (being able to list a project == being able to interact with its cache). I think it's fine as long as we don't forget it, which gets tricky as more work is happening, so I was wondering if we could add an introspection endpoint in the server that can return information about the token. By doing so the authorization contract is more explicit.